docs: add cvss info and suggest further procedural details (#419)
openedx/open-edx-proposals: oeps/processes/oep-0060-proc-sec-group.rst

The email will be added to SECURITY.md in Github, and an email link will be provided on openedx.org along with the Open edX
security policy.

Reports are triaged by the on-call security working group member
----------------------------------------------------------------
The on-call security working group member will triage the incoming report.

They will:

* Verify the report
* Reply to the reporter
* Forward operator-specific vulnerabilities to any relevant Open edX operators
* Score the severity of the disclosure
* Identify the affected repositories
* Open a GitHub repository security advisory for the vulnerability in the relevant repositories
* Notify the Open edX maintainers of those repositories of the vulnerability

> The security working group gives security severity and remediation advice to maintainers

Reviewer: ```suggestion
Issue Triage and Routing
```
These section titles are very verbose, here's one suggestion for a shorter one but if you have a different idea, feel free to use that also.

Author: Great suggestion.

----------------------------------------------------------------------------------------
The security working group is responsible for advising on security and providing security-scoring-as-a-service for maintainers.

Maintainers are encouraged and welcome to email security@openedx.org with any security questions they would like advice on. The security working group will do their best to provide good security advice.

The security working group will also score incoming vulnerabilities using the `Common Vulnerability Scoring System (CVSS), version 3.1`_. This system provides a way to score vulernabilities from 0.0 to 10.0, which map to a severity of Low, Medium, High, or Critical.

This numerical score and qualitative severity rating will be provided to maintainers so they are better informed on how severe the vulnerability is and have a sense of how important it may be to interrupt their work on the repository to resolve the vulnerability.

.. _Common Vulnerability Scoring System (CVSS), version 3.1: https://www.first.org/cvss/v3.1/specification-document

> Maintainers are ultimately responsible for the security of their repositories

Reviewer: ```suggestion
Maintainer Responsibilty
```

Author: Will commit.

-----------------------------------------------------------------------------

> Maintainers are responsible for resolving disclosed security vulnerabilies and notifying the security working group once they are resolved.

Reviewer: ```suggestion
Maintainers are ultimately responsible for resolving disclosed security vulnerabilities and notifying the security working group once they are resolved.
```

Author: Will commit.

To assist the maintainer with keeping track of their repository's vulnerabilities, the security working group will create a `GitHub repository security advisory`_ pre-assigned with the scored severity.

Maintainers are requested to notify the security working group if they would like to adjudicate a vulnerability to a different severity than what was originally triaged. The security working group must accept the maintainer's adjudication, but should comment on any considerations around the adjudication.

Maintainers will be reminded to remediate disclosures in proportion to the severity of the disclosure. The following table shows the reminder frequency until resolution of each severity classification:

========    =====   ==================
Severity    Score   Reminder frequency
========    =====   ==================

> Low         ≥0.1    Once a semester

Reviewer: What does semester mean in this context?  Is it meant to convey "twice annually"?

Author: Yes! I will amend.

Medium      ≥4.0    Once a quarter
High        ≥7.0    Once a month

> Critical    ≥9.0    Once a week

Reviewer: At first glance this seemed to infrequent but given that after the initial notification it's the maintainer's responsibility to prioritize and resolve the issue, I think it makes sense.

========    =====   ==================

The on-call security working group member is responsible for sending reminders for any vulnerability reminders due during the time they are on-call.

.. note::

    Maintainers are encouraged to reach out to security@openedx.org if they need any help remediating security issues.

.. _GitHub repository security advisory: https://docs.github.com/en/code-security/security-advisories/repository-security-advisories/creating-a-repository-security-advisory

> Details of security vulnerabilities are private until a fix is released

Reviewer: ```suggestion
Security Vulnerability Privacy
```

Author: Will commit.

-----------------------------------------------------------------------

> Members of the Open edX community with access to the details of a disclosure are asked to keep those details private until a fix is publically released for the vulnerability.

Reviewer: ```suggestion
Details of security vulnerabilities will remain private until a fix is released. Members of the Open edX community with access to the details of a disclosure are asked to keep those details private until a fix is publicly released for the vulnerability.
```

Author: Will commit.

Maintainers should use GitHub's `temporary private fork`_ feature within the GitHub repository security advisory created by the security working group to keep the implementation details of a fix for a vulnerability private.

.. _temporary private fork: https://docs.github.com/en/code-security/security-advisories/repository-security-advisories/collaborating-in-a-temporary-private-fork-to-resolve-a-repository-security-vulnerability

> Operators are ultimately responsible for the security of their Open edX instance

Reviewer: Any reason this shouldn't just go into the `Guidance for operators` section?

Author: You're right, most of the information is down there already. Removed.

--------------------------------------------------------------------------------
The security working group will put forth good faith effort to responsibly forward disclosures related to the operation of a specific Open edX instance to the relevant operator.

> Operators wishing to register a desired primary point of contact for security-related matters should email security@openedx.org with the name, domain, contact information for an individual point of contact for security, and an email address to which security reports should be forwarded.

Reviewer: ```suggestion
Operators wishing to register a desired primary point of contact for security-related matters should email security@openedx.org with the name, domain, contact information for an individual point of contact for security, and an email address to which security reports should be forwarded.  We recommend that you provide a consistent alias for this rather than the e-mail address of a specific person to keep the process resilient.  e.g. ``security@companyx.com``
```

Author: Since I deleted this section per your previous comment, I moved the spirit of this change to the Guidance for Operators section. Let me know what you think and please feel free to propose other edits.

Operators are encouraged to forward or to invite reporters to forward to security@openedx.org any security issue they receive that may be relevant to the Open edX project.

.. _security-releases:

> Maintainers will work together with the on-call security working group member to release the security fix

Reviewer: Why this change, the shorter title seems like it captures what this section is about.  Maybe something in between like `Vulnerability Release Process`

Author: Was trying to keep the parallelism with the long form of the first section title! Will rename back to `Security Releases`.

---------------------------------------------------------------------------------------------------------
The current process for releasing security fixes involves sending a disclosure and security patch to members of the
open source security email list and waiting two days before making the patch public.  Instead of this process, we
propose the following security release process:

#. The maintainer will email security@openedx.org notifying that a fix is ready to be merged and the date and time of when they would like to merge it. The merge time should be at least 72 hours after the security patch is announced.

> #. The on-call security working group member will create an announcement post in the `Security Announcement Section`_ on https://discuss.openedx.org. It should specify the affected repository, the date and time at which the patch will become public, and the severity of the vulnerability it fixes.

Reviewer: Gut reaction: This feels like something we could just ask the maintainer to do if we provide them a doc to follow, but I'm fine with it being a SWG responsibility if you feel strongly about it.

Author: Not at all, I prefer less work honestly. Amended.

   * How can we take advantage of the alerts that Github provides for security prioritization?
#. Iterations on the security process.
#. Review industry best practices that we should consider implementing.
#. Running an annual security survey.

Consequences
************

Security Working Group
======================

> Invite only

Reviewer: ```suggestion
Invite Only
```
Shouldn't titles be fully capitalized?  Using [this](https://www.grammarly.com/blog/capitalization-in-the-titles/) as my guide.

Author: Yes, was imitating the first title `Reports come into security@openedx.org`. Amended.

and will then be evaluated by the existing working group members. We hope to have six to seven members of the working
group at a time.

Security work will be a mix of private and public tickets
---------------------------------------------------------
Proactive work for security improvements will be made public, while vulnerability reports will be handled privately.

> We propose using Github Security Advisories to handle the triage of vulnerability reports to the owners of

Author: @feanil I think for GitHub security advisories to work for us, we will need to grant maintainers repository ownership / GitHub repository admin permissions on this repo. Is this in alignment with tCRIL's plans for maintainership?

Author: (What I am reading: https://docs.github.com/en/code-security/security-advisories/repository-security-advisories/permission-levels-for-repository-security-advisories)

Reviewer: @pshiu I think there are a couple of ways to solve this but at the very least I expect the security working group will need admin access to all the open source repos to do their job.  I think we're okay with this.

For the larger question of maintainers and admin,  I'm gonna think about it more but it's come up in a few other places as well so I think we'll probably add an ADR to the maintainers OEP soon.

Author: Always happy with more permissions, but there is a beta Security Manager role GitHub offers too that we might be able to tinker around with. [[docs](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/managing-security-managers-in-your-organization)]

Security work will be a mix of private and public tickets. Proactive work for security improvements will be made public, while vulnerability reports will be handled privately.

> We propose using GitHub Security Advisories to handle the triage of vulnerability reports to the owners of vulnerable components.

Reviewer: Not a problem to note here but I'm realizing that we'll need to keep a master list of draft advisories as a working group because github doesn't really provide a way to do that yet at the organization level.

and will then be evaluated by the existing working group members. We hope to have six to seven members of the working
group at a time.

> There will also be ways to participate in the security of the Open edX platform without being a member of the working group. Maintainers and core contributors of the Open edX community interested in volunteering to implement security hotfixes for other maintainers who may not have the bandwidth to immediately address vulnerabilities are encouraged to email their interest to security@openedx.org.

Author: @alangsto @feanil Wanted to draw your attention to this line that I added post-PR-creation to make sure you saw it and provide your thoughts!

We propose using Github Security Advisories to handle the triage of vulnerability reports to the owners of
vulnerable components.

> Member responsibilities

Reviewer: ```suggestion
Member Responsibilities
```

* If you believe this is a vulnerability in the Open edX Platform, please forward the issue to security@openedx.org
  or ask the reporter to re-file the issue to security@openedx.org

What will happen if a report is accidentally sent to security@openedx.org for the operation of my Open edX instance?

* Please let security@openedx.org know the best email to forward such reports to, along with your Open edX instance name, domain, and separate contact information for an indidual responsible for security at your organization. The security working group will do its best to forward such reports to the correct organization.

How do I receive notification of the release of upcoming security patches?

* Please watch the Open edX Discourse Security Announcements topic at https://discuss.openedx.org/c/announcements/security/19. If you are logged in, the button with a bell icon on the top right corner above the topic list will let you customize how you are notified of security announcements.

> Change History

Reviewer: Don't forget to add a section to the change history.

Author: Done.
